### PR TITLE
Add lora validations

### PR DESCRIPTION
**Pull request by aggarwal-k**

### CODE REVIEW

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    logger.info("Loading base model")
+    base_model = AutoModelForCausalLM.from_pretrained(base_model_path, device_map="auto", torch_dtype=torch.float16)
+    logger.info("Base model loaded")
+
+    logger.info("Loading lora adapters")
+    model = PeftModel.from_pretrained(base_model, lora_weights_path)
+    logger.info("lora model loaded")
+
+    input_ids = tokenizer(test_examples, return_tensors="pt").input_ids.to("cuda")
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:48]: why are we just taking the _input_ids_? why not the entire tokenized output?

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    "top_p": 0.7,
+    "temperature": 0.8
+}
+
+
+def validate_lora_weights(base_model_path: str, lora_weights_path: str, tokenizer_path: str, test_examples: list):
+    """Load model and make forward pass to validate lora weights."""
+    logger.info("Validating lora model weights")
+
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:38]: Given that the config is in a different folder, I think its best to pass config information while loading tokenzier and base model

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    logger.info("Loading base model")
+    base_model = AutoModelForCausalLM.from_pretrained(base_model_path, device_map="auto", torch_dtype=torch.float16)
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:41]: nit: we have to read finetune config params here to enable _trust_remote_code=true_ for instance

File: assets/training/finetune_acft_hf_nlp/components_v2/pipeline_components/text_generation/spec.yaml [hunk 4]
```diff
@@ -30,7 +30,7 @@ inputs:
   # Continual-Finetuning model path
   mlflow_model_path:
     type: mlflow_model
-    optional: true
+    optional: false
```
Comment (chaoqian) [inline assets/training/finetune_acft_hf_nlp/components_v2/pipeline_components/text_generation/spec.yaml:33]: why this change?

File: assets/training/finetune_acft_hf_nlp/components_v2/finetune/text_generation/spec.yaml [hunk 5]
```diff
@@ -15,7 +15,7 @@ code: ../../../src/
 inputs:
   mlflow_model_path:
     type: mlflow_model
-    optional: True
+    optional: False
```
Comment (chaoqian) [inline assets/training/finetune_acft_hf_nlp/components_v2/finetune/text_generation/spec.yaml:18]: do we need mlflow model to be required? should we allow pytorch model input too?

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        acft_custom_dimensions={
+            LoggingLiterals.PROJECT_NAME: PROJECT_NAME,
+            LoggingLiterals.PROJECT_VERSION_NUMBER: VERSION,
+            LoggingLiterals.COMPONENT_NAME: COMPONENT_NAME
+        },
+        azureml_pkg_denylist_logging_patterns=LOGS_TO_BE_FILTERED_IN_APPINSIGHTS,
+    )
+
+    print("validating lora weights")
+    validate_lora_weights(args.base_model_path, args.lora_weights_path, args.tokenizer_path, ["Hello"])
```
Comment (chaoqian) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:112]: looks like this validation just validate inference capability, but not quality: it cannot prevent inferencing garbage  output?

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            LoggingLiterals.COMPONENT_NAME: COMPONENT_NAME
+        },
+        azureml_pkg_denylist_logging_patterns=LOGS_TO_BE_FILTERED_IN_APPINSIGHTS,
+    )
+
+    # extract model and config path from mlflow base model path
+    base_model_path = os.path.join(args.mlflow_model_path, "data/model")
+    config_path = os.path.join(args.mlflow_model_path, "data/config")
+
+    validate_lora_weights(base_model_path, config_path, args.lora_weights_path, args.tokenizer_path, ["Hello"])
```
Comment (jpmann) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:118]: nit: we could have inferred tokenizer path as well from mlflow_model_path to be consistent with model and config paths

**Review by jpmann — APPROVED** [review]

File: assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    logger.info("Loading lora adapters")
+    model = PeftModel.from_pretrained(base_model, lora_weights_path)
+    logger.info("lora model loaded")
+
+    input_ids = tokenizer(test_examples, return_tensors="pt").input_ids.to("cuda")
+
+    try:
+        outputs = model.generate(inputs=input_ids, **GENERATION_CONFIG)
+        predictions_text = tokenizer.batch_decode(outputs)
+    except Exception:
```
Comment (amangupta26) [inline assets/training/finetune_acft_hf_nlp/src/validate_lora_weights/validate_lora_weights.py:56]: Can we log what was exception as well

**Review by amangupta26 — APPROVED** [review]

**Review by jpmann — APPROVED** [review]
